### ISSUE TITLE
Unclear guidance in ActiveRecord Callback docs

### ISSUE THREAD

**Issue by getaaron** [rails/rails#39418]
### Steps to reproduce
https://github.com/rails/rails/pull/38166 added this warning to the [ActiveRecord Callbacks docs](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks):

> WARNING. Care should be taken in callbacks to avoid updating attributes. For example, avoid running `update(attribute: "value")` and similar code during callbacks. This can alter the state of the model and may result in unexpected side effects during commit.

It recommends:

> Instead, you should try to assign values in the `before_create` or earlier callbacks.

However, the warning I just said said _not_ to update attributes during callbacks.

### Expected behavior

I expect to understand what code I should write when I want a callback to trigger an attribute change

### Actual behavior

This warning leaves me with a few questions:

- Is there a difference between "running `update(attribute: "value")`" ("during callbacks") and "assign[ing] values" (in the "`before_create` or earlier callbacks")? 
    - If yes, what's the difference between `update(attribute: "value")` and "assigning values"?
    - If no, we should rephrase to avoid ambiguity
- Are _some_ callbacks safe to call `update(attribute: "value")` in?
    - If yes, why does the warning just say "during callbacks"?
    - If no, how can I safely have a change to one value trigger an update to another?

Basically, I don't understand what to do in response to this warning.

## System configuration
**Rails version**: n/a, docs

**Ruby version**: n/a, docs

**Comment by getaaron** [comment]
@jules2689 @rafaelfranca hoping you can help with this since you worked on the original change. I'm happy to open a PR once I understand the intent of the warning. 🙃

**Comment by getaaron** [comment]
CC @eugeneius who is working on possibly related issue https://github.com/rails/rails/issues/39173

**Comment by jules2689** [comment]
Basically, by calling updates and saves in callbacks, it messes with internal state in ActiveRecord. Particularly it was this line:

https://github.com/rails/rails/blob/4a17d3e045ef059d5c82d1ade4f7eae37bc1e485/activerecord/lib/active_record/autosave_association.rb#L365-L369

This shows that a `before_save` handles how associations are saved. When you mess with that callback loop by inserting your own callbacks with `update`, that ivar gets set to false. This meant that your associations would not get saved if I remember correctly.

![Callback Flows(1)](https://user-images.githubusercontent.com/3074765/82824978-55256680-9e78-11ea-88fc-be2b61ce91f3.png)

To fix this, I added a `return ivar if defined?(ivar)` wrapper so that we only set it once.

This was then changed by @eugeneius in this PR (https://github.com/rails/rails/pull/39124) which explicitly makes auto-saving multiple times possible. I'm not saying this is the cause of the attached #39173, but it could be likely.

I'm not sure that that PR has the right approach as `||=` is a confusing assignment operator when dealing with booleans because it allows `false` to be overridden to `true` in some cases. This means that you might get multiple saves of associations. This is why I explicitly chose to use a `defined?` guard clause. I suspect we can validate what is happening by putting some logging statements or debuggers in that method and tracing what is going on.

**Comment by jules2689** [comment]
https://github.com/rails/rails/issues/39173#issuecomment-625157600

This comment pretty clearly outlines that they are defining an update in an `after_create` which is the basis of the issue I fixed in https://github.com/rails/rails/pull/38166

**Comment by jules2689** [comment]
In fact, reverting both my change and @eugeneius' change allows [this](https://github.com/rails/rails/issues/39173#issuecomment-625178248) to pass.

What is happening in this issue is that there are 2 callbacks:
```ruby
def before_save_collection_association
  @new_record_before_save ||= new_record?
end

def after_save_collection_association
  @new_record_before_save = false
end
```

By defining a callback in a callback, these are being confused. `after_save_collection_association` is supposed to set this new record to false, however you're calling a callback between the point this is set to `new` and before we can reset it to false. This is causing it to be called twice.

**Comment by getaaron** [comment]
@jules2689 Thanks, I'm still not sure about a trivial case though. For example, consider a `User` class with `sms_phone_number` (string) and `sms_number_is_verified` (boolean) model attributes. The expected behavior is that `sms_number_is_verified` reverts to `false` whenever `sms_phone_number` changes.

Will something like this work & is it safe? I see lots of examples of it in the wild but the warning seems to steer me away from it.

```ruby
class User < ApplicationRecord
  before_update :update_sms_verification

  def update_sms_verification
    if saved_change_to_sms_phone_number?
      sms_number_is_verified = false
    end
  end
end
```

The warning in the documentation is unclear here. I'm not calling `update(attribute: "value")` but I am changing an attribute inside a callback. Is this safe? Would it be safe in some callbacks but not others? Would calling `update(sms_number_is_verified: false)` be different than calling `sms_number_is_verified = false`?

**Comment by jules2689** [comment]
@getaaron yes, that should be safe. _Assigning_ attributes is ok, _updating_ or _saving_ attributes is not. In your example, you'd assign the variable and the normal save/update call will handle persisting the data to the database. That said, if you called `update(sms_number_is_verified: false)` you would be doing this:

1. `update`, `before_update` is called
2a) `before_update` calls `update`, go to (1) 
2b) `before_update` doesn't call `update` continue with the callbacks

In your example you'd go `1`, `2a`, `1`, `2b` (from the update in before_update), `2b` from the original callbacks. Whereas assigning would be `1`, `2b` (2a is never called).

Assigning attributes will simply change the value and not call save or any callbacks, while updating/saving in a callback literally hijacks the current callback stack (you're nesting callbacks in callbacks) which is bad because parts of Rails (implicitly at least) assumes we have full control over the stack (as we manipulate ivars).

In some cases, like the issue you linked, the hijacked callback stack causes issues because an ivar is manipulated multiple times and the rails callback stack doesn't handle all the various edge cases, so we get bugs. This is why Rails does not support callbacks in callbacks.

**Comment by getaaron** [comment]
@jules2689  Thanks that helps a lot! I opened a PR with some minor edits which I think will make the docs a bit clearer  https://github.com/rails/rails/pull/39433
